scripts/s3-support-bundles/s3_bundle_generate.sh

```diff
@@ -98,67 +100,105 @@ s3server_logdir=`cat $s3server_config | grep "S3_LOG_DIR:" | cut -f2 -d: | sed -
 authserver_logdir=`cat $authserver_config | grep "logFilePath=" | cut -f2 -d'=' | sed -e 's/^[ \t]*//' -e 's/#.*//' -e 's/^[ \t]*"\(.*\)"[ \t]*$/\1/'`
 backgrounddelete_logdir=`cat $backgrounddelete_config | grep "logger_directory:" | cut -f2 -d: | sed -e 's/^[ \t]*//' -e 's/#.*//' -e 's/^[ \t]*"\(.*\)"[ \t]*$/\1/'`
 
-# Compress each s3server core file present in /var/crash directory if available
-# these compressed core file will be available in /tmp/s3_support_bundle_<pid>/s3_core_files directory
-compress_core_files(){
-  core_filename_pattern="*/core-s3server.*"
-  for file in $s3_core_dir/*
-  do
-    if [[ -f "$file" && $file == $core_filename_pattern ]];
-    then
-        mkdir -p $s3_core_files
-        file_name=$(basename "$file")      # e.g core-s3server.234678
-        gzip -f -c $file > $s3_core_files/"$file_name".gz 2> /dev/null
-    fi
-  done
+# Collect call stack of latest <s3_core_files_max_count> s3server core files
+# from s3_core_dir directory, if available
+collect_core_files(){
+  core_filename_pattern="core-s3server.*.gz"
+  mkdir -p $s3_core_files
+  cwd=$(pwd)
+  cd $s3_core_dir
+  # get recent modified core files from directory
+  (ls -t $core_filename_pattern 2>/dev/null | head -$s3_core_files_max_count) | xargs -I '{}' cp '{}' $s3_core_files
+  cd $cwd
+  # check for empty directory for core files
+  if [ -z "$(ls -A $s3_core_files)" ];
+  then
+      rm -rf $s3_core_files
+  else
+      # iterate over the s3_core_files directory and extract call stack for each file
+      for s3corefile in "$s3_core_files"/*
+      do
+          zipped_core_name=$(basename "$s3corefile")
+          core_name=${zipped_core_name%".gz"}
+          callstack_file="$s3_core_files"/callstack."$core_name".txt
+
+          # unzip the core file, to read it using gdb
+          tar -xzf $s3corefile -C "$s3_core_files"
+
+          printf "Core name: $core_name\n" >> "$callstack_file"
+          printf "Callstack:\n\n" >> "$callstack_file"
+
+          # generate gdb bt and append into the callstack_file
+          gdb --batch --quiet -ex "thread apply all bt full" -ex "quit" $s3server_binary\
```

Comment:
For big core files, use gzip -l, figure out the 'inflated' size and only inflate if < 10GB , else just make a placeholder file

Reply:
done